snuba_sdk/query.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, replace
+from typing import List, Optional, Union
+
+from snuba_sdk import Expression
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import (
+    Column,
+    Condition,
+    Function,
+    Granularity,
+    Limit,
+    Offset,
+)
+
+
+@dataclass(frozen=True)
+class Query(Expression):
+    dataset: str
+    match: Entity
+    select: Optional[List[Union[Column, Function]]] = None
+    groupby: Optional[List[Union[Column, Function]]] = None
+    where: Optional[List[Condition]] = None
+    limit: Optional[Limit] = None
+    offset: Optional[Offset] = None
+    granularity: Optional[Granularity] = None
+
+    def set_match(self, match: Entity) -> Query:
+        new = replace(self, match=match)
+        new.validate()
+        return new
+
+    def set_select(self, select: List[Union[Column, Function]]) -> Query:
+        new = replace(self, select=select)
+        new.validate()
+        return new
+
+    def set_groupby(self, groupby: List[Union[Column, Function]]) -> Query:
+        new = replace(self, groupby=groupby)
+        new.validate()
+        return new
+
+    def set_where(self, conditions: List[Condition]) -> Query:
+        new = replace(self, conditions=conditions)
+        new.validate()
+        return new
+
+    def set_limit(self, limit: int) -> Query:
+        new = replace(self, limit=Limit(limit))
+        new.validate()
+        return new
+
+    def set_offset(self, offset: int) -> Query:
+        new = replace(self, offset=Offset(offset))
+        new.validate()
+        return new
+
+    def set_granularity(self, granularity: int) -> Query:
+        new = replace(self, granularity=Granularity(granularity))
+        new.validate()
+        return new
+
+    def validate(self) -> None:
+        # TODO: Contextual validation. E.g. top level functions in select
+        # require aliases
+        self.match.validate()
+        clauses = [self.select, self.groupby, self.where]
+        for exps in clauses:
+            if exps is not None:
+                for exp in exps:
+                    exp.validate()
+
+        self.limit.validate() if self.limit is not None else None
+        self.offset.validate() if self.offset is not None else None
+        self.granularity.validate() if self.granularity is not None else None
+
+    def translate(self) -> str:
```

Comment:
If you were talking about this when you were considering using a visitor, strong +1 on that. 
That would allow you separate the translation from the query structure itself, which means you can swap different translators (imagine a pretty print one with a simpler one) without bloating the structure of your query and all the expressions. This can blow up quickly.
Another use case could be allowing you to have a specific serializer for logs that does not serialize literals so identical queries with different project and time range are serialized in the same way.

Reply:
Yeah OK. I'll have to add that in. Even with just these two cases (validate and translate) it feels cumbersome.